### PR TITLE
Add check for developer roles to Harvest gateway with tests

### PR DESCRIPTION
**Pull request by RobertM905**

### CODE REVIEW

File: CryptoTechReminderSystem.Test/Gateway/HarvestGatewayTests.cs [hunk 1]
```diff
@@ -213,6 +227,8 @@ public void CanSendRequestOnceAtATime()
                
                 _harvestApi.ReceivedRequests.Count.Should().Be(1);
             }
+            
+
```
Comment (dushan-madetech) [inline CryptoTechReminderSystem.Test/Gateway/HarvestGatewayTests.cs]:
```suggestion
```

File: CryptoTechReminderSystem/Gateway/HarvestGateway.cs [hunk 2]
```diff
@@ -59,6 +59,21 @@ public IList<HarvestDeveloper> RetrieveDevelopers()
...
+            return (userRole.Contains("Software Engineer") ||
+                    userRole.Contains("Senior Software Engineer") ||
+                    userRole.Contains("Senior Engineer") ||
+                    userRole.Contains("Lead Engineer") ||
+                    userRole.Contains("Delivery Manager") ||
+                    userRole.Contains("SRE") ||
+                    userRole.Contains("Consultant") ||
+                    userRole.Contains("Delivery Principal")
+                   );
+
```
Comment (RobertM905) [inline CryptoTechReminderSystem/Gateway/HarvestGateway.cs]: Rogue blank line.

Comment (dushan-madetech) [reply CryptoTechReminderSystem/Gateway/HarvestGateway.cs]:
```suggestion
```

File: CryptoTechReminderSystem.Test/Gateway/HarvestGatewayTests.cs [hunk 3]
```diff
@@ -39,7 +39,7 @@ public void TearDown()
             private void SetUpUsersApiEndpoint(string id, string firstName, string lastName, string email)
             {
                 var json = $"{{  \"users\":[    {{      \"id\":{id},      \"first_name\":\"{firstName}\"" +
-                           $",      \"last_name\":\"{lastName}\",       \"email\":\"{email}\" , \"is_active\":\"true\"   }}  ]}}";
+                           $",      \"last_name\":\"{lastName}\",       \"email\":\"{email}\" , \"is_active\":\"true\", \"roles\": [\"Delivery Principal\"]  }}  ]}}";
```
Comment (C-gyorfi) [inline CryptoTechReminderSystem.Test/Gateway/HarvestGatewayTests.cs]: This my be a cleaner way to set up api response
`new { last_name = lastName, email = email, is_active = true, etc... }`

File: CryptoTechReminderSystem/Gateway/HarvestGateway.cs [hunk 4]
```diff
@@ -59,6 +59,21 @@ public IList<HarvestDeveloper> RetrieveDevelopers()
                 ).ToList();
         }
 
+        private bool CheckUserRole(JToken user)
+        {
+            var userRole = user["roles"].Select(role => role.ToString());
```
Comment (montashimo) [inline CryptoTechReminderSystem/Gateway/HarvestGateway.cs]: 😢

File: CryptoTechReminderSystem/Gateway/HarvestGateway.cs [hunk 5]
```diff
@@ -48,7 +48,7 @@ public IList<HarvestDeveloper> RetrieveDevelopers()
             var apiResponse = response.Result;
             var users = apiResponse["users"];
            
-            return users.Where(user => (bool)user["is_active"] != false)
+            return users.Where (user => (bool)user["is_active"] != false && CheckUserRole(user))
```
Comment (chubberlisk) [inline CryptoTechReminderSystem/Gateway/HarvestGateway.cs]:
```suggestion
            return users.Where (user => (bool)user["is_active"] != false && IsDeveloper(user))
```
If we decide to keep this method, I think we should rename it to something more readable.

File: CryptoTechReminderSystem/Gateway/HarvestGateway.cs (hunk 4 above)
Comment (RobertM905) [reply CryptoTechReminderSystem/Gateway/HarvestGateway.cs]: Not sure what you mean here?

File: CryptoTechReminderSystem.Test/Gateway/HarvestGatewayTests.cs (hunk 1 above)
Comment (RobertM905) [reply CryptoTechReminderSystem.Test/Gateway/HarvestGatewayTests.cs]: resolved in https://github.com/chubberlisk/crypto_tech/pull/30/commits/bf82ffb31fdb96302d5644d8613e869b7814018c

File: CryptoTechReminderSystem.Test/Gateway/HarvestGatewayTests.cs [hunk 6]
```diff
@@ -108,7 +108,21 @@ public void CanOnlyGetDevelopersWhoAreActive()
...
                 var response = _harvestGateway.RetrieveDevelopers();
                 response.First().FirstName.Should().Be("Dick");
-                response.Count.Should().Be(2);
+                response.Count.Should().Be(6);
+            }
+
+            [Test]
+            public void CanGeDevelopersOnly()
+            {
+                var jsonWithIsActive = File.ReadAllText(
```
Comment (Rumaysa) [inline CryptoTechReminderSystem.Test/Gateway/HarvestGatewayTests.cs]: Could probably rename this to json.

File: CryptoTechReminderSystem/Gateway/HarvestGateway.cs (hunk 4 above)
Comment (montashimo) [reply CryptoTechReminderSystem/Gateway/HarvestGateway.cs]: Lots of userRole.contains here - not very scalable

File: CryptoTechReminderSystem/Gateway/HarvestGateway.cs (hunk 5 above)
Comment (C-gyorfi) [reply CryptoTechReminderSystem/Gateway/HarvestGateway.cs]:
```suggestion
            return users.Where (user => !((bool)user["is_active"]) && CheckUserRole(user))
```
😉

File: CryptoTechReminderSystem.Test/Gateway/HarvestGatewayTests.cs [hunk 7]
```diff
@@ -39,7 +39,7 @@ public void TearDown()
             private void SetUpUsersApiEndpoint(string id, string firstName, string lastName, string email)
             {
                 var json = $"{{  \"users\":[    {{      \"id\":{id},      \"first_name\":\"{firstName}\"" +
```
Comment (Rumaysa) [inline CryptoTechReminderSystem.Test/Gateway/HarvestGatewayTests.cs]: is there a need for this as we have a json file now?

File: CryptoTechReminderSystem/Gateway/HarvestGateway.cs (hunk 5 above)
Comment (RobertM905) [reply CryptoTechReminderSystem/Gateway/HarvestGateway.cs]: applied the fix to https://github.com/chubberlisk/crypto_tech/pull/30/commits/55f45f507bf7296b62455a671f2ab24d79c75f36 for now

File: CryptoTechReminderSystem/Gateway/HarvestGateway.cs [hunk 8]
```diff
@@ -48,7 +48,7 @@ public IList<HarvestDeveloper> RetrieveDevelopers()
             var apiResponse = response.Result;
             var users = apiResponse["users"];
            
-            return users.Where(user => (bool)user["is_active"] != false)
+            return users.Where (user => (bool)user["is_active"] != false && IsDeveloper(user))
```
Comment (C-gyorfi) [inline CryptoTechReminderSystem/Gateway/HarvestGateway.cs]:
```suggestion
            return users.Where (user => (bool)user["is_active"]  && IsDeveloper(user))
```

File: CryptoTechReminderSystem/Gateway/HarvestGateway.cs [hunk 9]
```diff
@@ -48,7 +48,7 @@ public IList<HarvestDeveloper> RetrieveDevelopers()
             var apiResponse = response.Result;
             var users = apiResponse["users"];
            
-            return users.Where(user => (bool)user["is_active"] != false)
+            return users.Where (user => (bool)user["is_active"]  && IsDeveloper(user))
                 .Select(developer => new HarvestDeveloper()
```
Comment (C-gyorfi) [inline CryptoTechReminderSystem/Gateway/HarvestGateway.cs]:
```suggestion
                .Select(developer => new HarvestDeveloper
```

File: CryptoTechReminderSystem.Test/Gateway/HarvestGatewayTests.cs (hunk 6 above)
Comment (RobertM905) [reply CryptoTechReminderSystem.Test/Gateway/HarvestGatewayTests.cs]: removed this test in https://github.com/chubberlisk/crypto_tech/pull/30/commits/997365c3fbec53b1823542213cac8b80db7f17fb

File: CryptoTechReminderSystem.Test/Gateway/HarvestGatewayTests.cs (hunk 7 above)
Comment (RobertM905) [reply CryptoTechReminderSystem.Test/Gateway/HarvestGatewayTests.cs]: will be sorted in a separate PR

File: CryptoTechReminderSystem.Test/Gateway/HarvestGatewayTests.cs (hunk 3 above)
Comment (RobertM905) [reply CryptoTechReminderSystem.Test/Gateway/HarvestGatewayTests.cs]: Will be sorted in a separate PR

File: CryptoTechReminderSystem/Gateway/HarvestGateway.cs (hunk 4 above)
Comment (RobertM905) [reply CryptoTechReminderSystem/Gateway/HarvestGateway.cs]: We've refactored the `IsDeveloper` method in this commit https://github.com/chubberlisk/crypto_tech/pull/30/commits/718368640db4514a307d2a941125490a15b42f7b. What do you think?

Comment (RobertM905) [reply CryptoTechReminderSystem/Gateway/HarvestGateway.cs]: @montashimo

File: CryptoTechReminderSystem/Gateway/HarvestGateway.cs [hunk 10]
```diff
@@ -47,16 +58,21 @@ public IList<HarvestDeveloper> RetrieveDevelopers()
...
+                    Id = (int) developer["id"],
+                    FirstName = developer["first_name"].ToString(),
+                    LastName = developer["last_name"].ToString(),
+                    Email = developer["email"].ToString()
+                }
+            ).ToList();
+        }
+        
+        private bool IsDeveloper(JToken user)
+        {
```
Comment (montashimo) [inline CryptoTechReminderSystem/Gateway/HarvestGateway.cs]: What will happen here if there's no roles array? Is that a possibility?

File: CryptoTechReminderSystem/Gateway/HarvestGateway.cs [hunk 11]
```diff
@@ -15,6 +15,17 @@ public class HarvestGateway : IHarvestDeveloperRetriever, ITimeSheetRetriever
         private readonly string _token;
         private readonly string _accountId;
         private readonly string _userAgent;
+        private readonly List<string> _listOfDeveloperRoles = new List<string>
```
Comment (montashimo) [inline CryptoTechReminderSystem/Gateway/HarvestGateway.cs]: Could this be an environment variable?

File: CryptoTechReminderSystem/Gateway/HarvestGateway.cs (hunk 4 above)
Comment (montashimo) [reply CryptoTechReminderSystem/Gateway/HarvestGateway.cs]: These changes look good. Have made a couple of other suggestions 👍

File: CryptoTechReminderSystem/Gateway/HarvestGateway.cs (hunk 10 above)
Comment (yusufsheiqh) [reply CryptoTechReminderSystem/Gateway/HarvestGateway.cs]: I believe Harvest has an empty `roles` array if a person has no roles.

Comment (montashimo) [reply CryptoTechReminderSystem/Gateway/HarvestGateway.cs]: OK - good to know.

File: CryptoTechReminderSystem/Gateway/HarvestGateway.cs (hunk 11 above)
Comment (RobertM905) [reply CryptoTechReminderSystem/Gateway/HarvestGateway.cs]: For this PR this is probably unnecessary - we will do this as a separate task.

Comment (RobertM905) [reply CryptoTechReminderSystem/Gateway/HarvestGateway.cs]: We'll be working on this next and putting it into a separate PR. We'll send you a message when it's ready.

**Pull request merged**